### PR TITLE
Support for string parser

### LINKED ISSUE #51: Load from string?
Was this discussed before? i couldn't find an issue about it. 

Loading an XML from string will be beneficial since sometimes we don't have a file saved on the file system to load.

### LINKED ISSUE #108: Add support to read from string
Add support for reading from strings and [writing to](https://github.com/hassankhan/config/issues/107) strings instead of files. This could be done with methods` fromString($format, $data)` and `toString($format, $data)`.

Note that the format parameter should be case insensitive.

**Example of reading:**
```php
$settings = <<<FOOBAR
{
  "application": {
    "name": "configuration",
    "secret": "s3cr3t"
  },
  "host": "localhost",
  "port": 80,
  "servers": [
    "host1",
    "host2",
    "host3"
  ]
}
FOOBAR;
$conf = Config::fromString('JSON', $json);
```

**Example of writing:**
```php
$json = $conf->toString('JSON')
$conf->toString('YAML', $yaml);
```

### PR DESCRIPTION
**Pull request by filips123**
It add support for parsing strings and specifying file and string parsers. It closes #108 and closes #51. It also closes #109.
It also fixes some styling and tests.

Documentation is also updated.

### CODE REVIEW

**Comment by hassankhan** [comment]
Looks good to me overall, just waiting on @DavidePastore's review 👍

**Comment by filips123** [comment]
I've fixed documentation and added warning about arbitrary code execution.

**Comment by hassankhan** [comment]
Thanks for all your work @filips123 👍 Does this PR supersede #109?

**Comment by filips123** [comment]
Yes.

**Comment by hassankhan** [comment]
I've had another look at the PR - I guess my main concern is that there seems to be some code duplication between the `StringParser` and `FileParser` classes. Could we perhaps create an `AbstractParser` with similar abstract classes that both the `*Parser` classes then extend?

**Comment by filips123** [comment]
I've been thinking about this, but I don't know how exactly should I make this. Do you have some ideas?
I will probably add support for writing configuration to files and strings in the future, so this should also be considered.

File: src/StringParser/Ini.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace Noodlehaus\StringParser;
+
+use Noodlehaus\Exception\ParseException;
+
+/**
+ * INI string parser
+ *
+ * @package    Config
+ * @author     Jesus A. Domingo <jesus.domingo@gmail.com>
```
Comment (DavidePastore) [inline src/StringParser/Ini.php]: Please, add yourself as author of this class, removing the others.

File: src/StringParser/AbstractStringParser.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Noodlehaus\StringParser;
+
+/**
+ * Abstract string parser
+ *
+ * @package    Config
+ * @author     Jesus A. Domingo <jesus.domingo@gmail.com>
```
Comment (DavidePastore) [inline src/StringParser/AbstractStringParser.php]: Please, add yourself as author of this class, removing the others.

File: src/StringParser/Json.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace Noodlehaus\StringParser;
+
+use Noodlehaus\Exception\ParseException;
+
+/**
+ * JSON string parser
+ *
+ * @package    Config
+ * @author     Jesus A. Domingo <jesus.domingo@gmail.com>
```
Comment (DavidePastore) [inline src/StringParser/Json.php]: Please, add yourself as author of this class, removing the others.

File: src/StringParser/Php.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+use Exception;
+use Noodlehaus\Exception\ParseException;
+use Noodlehaus\Exception\UnsupportedFormatException;
+
+/**
+ * PHP string parser
+ *
+ * @package    Config
+ * @author     Jesus A. Domingo <jesus.domingo@gmail.com>
```
Comment (DavidePastore) [inline src/StringParser/Php.php]: Please, add yourself as author of this class, removing the others.

File: src/StringParser/StringParserInterface.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace Noodlehaus\StringParser;
+
+/**
+ * Config string parser interface
+ *
+ * @package    Config
+ * @author     Jesus A. Domingo <jesus.domingo@gmail.com>
```
Comment (DavidePastore) [inline src/StringParser/StringParserInterface.php]: Please, add yourself as author of this class, removing the others.

File: src/StringParser/Xml.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace Noodlehaus\StringParser;
+
+use Noodlehaus\Exception\ParseException;
+
+/**
+ * XML string parser
+ *
+ * @package    Config
+ * @author     Jesus A. Domingo <jesus.domingo@gmail.com>
```
Comment (DavidePastore) [inline src/StringParser/Xml.php]: Please, add yourself as author of this class, removing the others.

File: src/StringParser/Yaml.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+use Exception;
+use Symfony\Component\Yaml\Yaml as YamlParser;
+use Noodlehaus\Exception\ParseException;
+
+/**
+ * YAML string parser
+ *
+ * @package    Config
+ * @author     Jesus A. Domingo <jesus.domingo@gmail.com>
```
Comment (DavidePastore) [inline src/StringParser/Yaml.php]: Please, add yourself as author of this class, removing the others.

File: src/Config.php [hunk 8]
```diff
@@ -5,9 +5,12 @@
 use Noodlehaus\Exception\FileNotFoundException;
 use Noodlehaus\Exception\UnsupportedFormatException;
 use Noodlehaus\Exception\EmptyDirectoryException;
+use InvalidArgumentException;
+use Noodlehaus\FileParser\FileParserInterface;
+use Noodlehaus\StringParser\StringParserInterface;
 
 /**
- * Config
+ * Configuration reader and writer for PHP.
```
Comment (DavidePastore) [inline src/Config.php]: Please, add yourself as author in every file you edited.

File: src/StringParser/Ini.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
+<?php
```
Comment (DavidePastore) [inline src/StringParser/Ini.php]: This file is very similar to FileParser/Ini.php. We should share the similar functionalities in a common place. The same issue is present in Json, Xml, Yaml and in part in Php parsers.

**Review by DavidePastore — CHANGES_REQUESTED** [review]
@filips123 Thanks for all your work! 😮 Please, improve the code following @hassankhan feedback and my code reviews.

**Comment by hassankhan** [comment]
@filips123, with regards to "write" support, do you mean writing config values back to a file, or are you talking about adding methods like `update()` and `delete()`?

**Comment by filips123** [comment]
I currently can't program until next Thursday so I will add changes then.

**Comment by filips123** [comment]
@hassankhan I mean writing to files and also methods like `update()` and `delete()`.

File: src/StringParser/Ini.php (hunk 9 above)
Comment (filips123) [reply src/StringParser/Ini.php]: I've been thinking about this, but I don't know how exactly should I make this. Do you have some ideas?

**Comment by hassankhan** [comment]
Hi @filips123, thanks for replying. No worries, take your time 😄.

So I'd just like to mention that we've discussed adding functionality similar to this (#23 and #76 ) and decided against it in the end, as we do not feel that configuration values should be changed during application runtime.

Writing to files, however, is something that could definitely be handy for a variety of use-cases (concatenating various configs and writing the result to a file via a PHP CLI script).

**Comment by filips123** [comment]
Writing to files (or strings) would also be usable for user friendly installation programs.

File: src/Config.php (hunk 8 above)
Comment (filips123) [reply src/Config.php]: I've fixed this.

File: src/StringParser/Ini.php (hunk 1 above)
Comment (filips123) [reply src/StringParser/Ini.php]: I've fixed this.

File: src/StringParser/AbstractStringParser.php (hunk 2 above)
Comment (filips123) [reply src/StringParser/AbstractStringParser.php]: I've fixed this.

File: src/StringParser/Json.php (hunk 3 above)
Comment (filips123) [reply src/StringParser/Json.php]: I've fixed this.

File: src/StringParser/Php.php (hunk 4 above)
Comment (filips123) [reply src/StringParser/Php.php]: I've fixed this.

File: src/StringParser/StringParserInterface.php (hunk 5 above)
Comment (filips123) [reply src/StringParser/StringParserInterface.php]: I've fixed this.

File: src/StringParser/Xml.php (hunk 6 above)
Comment (filips123) [reply src/StringParser/Xml.php]: I've fixed this.

File: src/StringParser/Yaml.php (hunk 7 above)
Comment (filips123) [reply src/StringParser/Yaml.php]: I've fixed this.

**Comment by filips123** [comment]
@hassankhan @DavidePastore I've fixed some requested changes about authors, but I still don't know how to join file and string parser. Do you have some ideas?

**Comment by hassankhan** [comment]
I think the best way would be as I mentioned [here](https://github.com/hassankhan/config/pull/111#issuecomment-407459861). It also may be worthwhile renaming any  `*Parser` classes to `*Reader` instead, which will simplify things when the inevitable '*Writer` classes are added.

**Comment by filips123** [comment]
@hassankhan What function should be joined and how exactly? Maybe we could just have string parser and then pass file content to them if needed?

**Comment by filips123** [comment]
@hassankhan @DavidePastore I've joined file and string parsers.

**Review by DavidePastore — APPROVED** [review]

**Comment by filips123** [comment]
I also updated PHPUnit to namespaced syntax and added PHP 7.2 to Travis CI (see #110).

**Comment by filips123** [comment]
Can you please quickly merge this PR because I want to start working on configuration writing?

**Comment by hassankhan** [comment]
Really sorry for the delay, is there any chance you could rebase your PR on the `develop` branch? If you could do that, I'll merge the PR immediately :+1:

**Comment by filips123** [comment]
I changed branch.

**Review by hassankhan — APPROVED** [review]

**Comment by hassankhan** [comment]
Just waiting for the build to pass (expect it should be fine), will tag a new version shortly after.

**Comment by hassankhan** [comment]
Aaaaaand released as 2.0.0 🎉 !! 

Thanks @filips123 for your contribution, much appreciated 👍

**Pull request merged**